hibernate-core/src/main/java/org/hibernate/sql/ast/tree/expression/ColumnReference.java

```diff
@@ -25,32 +25,43 @@
  * Models a reference to a Column in a SQL AST
  *
  * @author Steve Ebersole
+ * @author Nathan Xu
  */
 public class ColumnReference implements Expression, Assignable {
 	private final String qualifier;
 	private final String columnExpression;
+	private final boolean isColumnExpressionFormula;
 	private final String referenceExpression;
 	private final JdbcMapping jdbcMapping;
 
 	public ColumnReference(
 			String qualifier,
 			String columnExpression,
+			boolean isColumnExpressionFormula,
 			JdbcMapping jdbcMapping,
 			SessionFactoryImplementor sessionFactory) {
 		this.qualifier = StringHelper.nullIfEmpty( qualifier );
-		this.columnExpression = columnExpression;
-		this.referenceExpression = this.qualifier == null
-				? columnExpression
-				: this.qualifier + "." + columnExpression;
+		if ( isColumnExpressionFormula ) {
+			assert qualifier != null;
+			this.columnExpression = StringHelper.replace( columnExpression, Template.TEMPLATE, qualifier );
+		}
+		else {
+			this.columnExpression = columnExpression;
+		}
+		this.isColumnExpressionFormula = isColumnExpressionFormula;
+		this.referenceExpression = this.qualifier == null || isColumnExpressionFormula
+				? this.columnExpression
+				: this.qualifier + "." + this.columnExpression;
 		this.jdbcMapping = jdbcMapping;
 	}
```

Comment:
This class change should be the facade of the code changes in this PR and starting with it is a good idea to grasp this seemingly big PR.